### PR TITLE
Pilot 1870: update the new restriction rules

### PR DESCRIPTION
**Pull request by colorzzr**
## Summary

- block the upload action to core in local env.
- block cross project data manipulation
- add new proxy for download api
- review the zone-based policy
- remove the redundant auth check

## JIRA Issues

Pilot 1870

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Refactor or reformatting

## Testing

Are there any new or updated tests to validate the changes?

- [x] Yes
- [ ] No

## Test Directions

test new proxy api and auth logic

### CODE REVIEW

File: app/routers/v1/api_project.py [hunk 1]
```diff
@@ -80,35 +85,34 @@ async def list_project(self, page=0, page_size=10, order='created_at', order_by=
...
-        val_result, val_error = validate_upload_event(
-            data.zone, self.current_identity, project_code
+
+        # check role based permission within project
+        permission = await has_permission(
+            self.current_identity,
+            project_code,
+            'file',
+            data.zone,
+            'view',
```
Comment (gregmccoy) [inline app/routers/v1/api_project.py:105]: This should probably be the `upload` permission

File: app/routers/v1/api_project.py [hunk 2]
```diff
@@ -137,18 +143,78 @@ async def project_file_preupload(
...
+            {self.current_identity}'
+        )
+
+        # check role based permission within project
+        permission = await has_permission(
+            self.current_identity,
+            project_code,
+            'file',
+            data.zone,
+            'view',
```
Comment (gregmccoy) [inline app/routers/v1/api_project.py:187]: This should probably be `download` permission

File: app/routers/v1/api_project.py [hunk 3]
```diff
@@ -137,18 +143,78 @@ async def project_file_preupload(
...
+        )
+
+        if not permission:
+            error_msg = f'Unauthorized upload action on project {project_code}'
```
Comment (gregmccoy) [inline app/routers/v1/api_project.py:191]: Error should say download

File: app/routers/v1/api_validation.py [hunk 4]
```diff
@@ -142,6 +167,8 @@ async def validate_env(self, request_payload: EnvValidatePost):
                 return api_response.json_response()
         else:
             current_zone = ConfigClass.CORE_ZONE_LABEL.lower()
+        # current_zone = vm_info.get("zone")
```
Comment (gregmccoy) [inline app/routers/v1/api_validation.py:170]: please remove

File: pyproject.toml [hunk 5]
```diff
@@ -6,28 +6,26 @@ authors = ["Indoc Research"]
...
 asyncpg = "0.25.0"
-psycopg2-binary = "2.8.5"
 opentelemetry-exporter-jaeger = "1.6.2"
 opentelemetry-instrumentation = "0.26b1"
 opentelemetry-instrumentation-fastapi = "0.26b1"
 opentelemetry-instrumentation-sqlalchemy = "0.26b1"
 opentelemetry-instrumentation-httpx = "0.26b1"
 opentelemetry-instrumentation-asyncpg = "0.26b1"
 fastapi-health = "^0.4.0"
+pre-commit = "^2.20.0"
```
Comment (gregmccoy) [inline pyproject.toml:27]: This should be a dev dependency

File: app/resources/authorization/decorator.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    # check the ip from signed payload must same as request
+    if vm_info.get('ip') != incoming_ip:
+        raise SourceIpMismatched('The ip of VM does not matched with source ip')
+
+    # check the project should be allowed within range
+    elif project_code not in vm_info.get('project_code'):
+        raise ProjectCodeMismacthed('The project of VM does not matched with query')
+
+    return
```
Comment (gregmccoy) [inline app/resources/authorization/decorator.py:104]: Can we remove the empty return, having no return still returns None.

File: app/resources/authorization/decorator.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    permit_action = restrict_zone.get(current_zone)
+    permit_zone = permit_action.get(action)
+    if target_zone not in permit_zone:
+        attempt = 'upload to' if action == ValidAction.UPLOAD else 'download from'
+        error = f'Invalid action: {attempt} {target_zone} in {current_zone}'
+        raise InvalidAction(error)
+
+    return
```
Comment (gregmccoy) [inline app/resources/authorization/decorator.py:134]: Same here

**Review by gregmccoy — CHANGES_REQUESTED** [review]

File: app/resources/authorization/decorator.py (hunk 7 above)
Comment (colorzzr) [reply app/resources/authorization/decorator.py:134]: sure I will update them now. thanks!

**Review by gregmccoy — APPROVED** [review]

**Pull request merged**